### PR TITLE
Flink: Enrich flink actions

### PR DESCRIPTION
**Pull request by Reo-LEI**
Currently, flink actions only have `RewriteDataFilesAction`, and I plan to implement `expireSnapshots`, `rewriteManifests` and `deleteOrphanFiles` actions for flink. I will refer to `SparkActions` implementation and trying transplant them to flink.

### CODE REVIEW

**Comment by Reo-LEI** [comment]
Now, the `expireSnapshots` coding is finish, but I'm blocking by a Kyro Serialization Error when I testing.

Kyro raise NEP when serialize `hadoopConf` of `HadoopFileIO` when flink create `CombinedScanTask` source. the full stack as follow:
```
Exception in thread "main" java.lang.RuntimeException: Serializing the source elements failed: java.lang.NullPointerException
Serialization trace:
hadoopConf (org.apache.iceberg.hadoop.HadoopFileIO)
io (org.apache.iceberg.AllManifestsTable$ManifestListReadTask)
tasks (org.apache.iceberg.BaseCombinedScanTask)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.fromCollection(StreamExecutionEnvironment.java:1046)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.fromCollection(StreamExecutionEnvironment.java:1021)
	at org.apache.iceberg.flink.source.MetadataTableSource$Builder.build(MetadataTableSource.java:92)
	at org.apache.iceberg.flink.actions.BaseFlinkAction.loadMetadataTable(BaseFlinkAction.java:117)
	at org.apache.iceberg.flink.actions.BaseFlinkAction.buildValidDataFileTable(BaseFlinkAction.java:92)
	at org.apache.iceberg.flink.actions.ExpireSnapshotsAction.buildValidFileTable(ExpireSnapshotsAction.java:122)
	at org.apache.iceberg.flink.actions.ExpireSnapshotsAction.expire(ExpireSnapshotsAction.java:139)
	at org.apache.iceberg.flink.actions.ExpireSnapshotsAction.execute(ExpireSnapshotsAction.java:117)
	at org.apache.iceberg.flink.actions.ExpireSnapshotsAction.execute(ExpireSnapshotsAction.java:44)
	at cn.walden.stream.IcebergFlinkActions.main(IcebergFlinkActions.java:28)
Caused by: java.io.IOException: Serializing the source elements failed: java.lang.NullPointerException
Serialization trace:
hadoopConf (org.apache.iceberg.hadoop.HadoopFileIO)
io (org.apache.iceberg.AllManifestsTable$ManifestListReadTask)
tasks (org.apache.iceberg.BaseCombinedScanTask)
	at org.apache.flink.streaming.api.functions.source.FromElementsFunction.<init>(FromElementsFunction.java:93)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.fromCollection(StreamExecutionEnvironment.java:1044)
	... 9 more
Caused by: com.esotericsoftware.kryo.KryoException: java.lang.NullPointerException
Serialization trace:
hadoopConf (org.apache.iceberg.hadoop.HadoopFileIO)
io (org.apache.iceberg.AllManifestsTable$ManifestListReadTask)
tasks (org.apache.iceberg.BaseCombinedScanTask)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:82)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:495)
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:523)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:61)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:495)
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:599)
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.write(DefaultArraySerializers.java:348)
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.write(DefaultArraySerializers.java:289)
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:523)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:61)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:495)
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:599)
	at org.apache.flink.api.java.typeutils.runtime.kryo.KryoSerializer.serialize(KryoSerializer.java:316)
	at org.apache.flink.streaming.api.functions.source.FromElementsFunction.<init>(FromElementsFunction.java:89)
	... 10 more
Caused by: java.lang.NullPointerException
	at com.esotericsoftware.kryo.util.DefaultClassResolver.writeClass(DefaultClassResolver.java:80)
	at com.esotericsoftware.kryo.Kryo.writeClass(Kryo.java:488)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:57)
	... 23 more
```

I'm trying to figure out the reason of that, but I have been blocking for this for a while and I don't have a clue. Could someone give me a little help? @rdblue @aokolnychyi @openinx

**Comment by rdblue** [comment]
Looks like the FileIO was not prepared to be serialized by wrapping its Hadoop Configuration in SerializableConfiguration.

**Comment by kbendick** [comment]
I have _potentially_ come across a similar issue when using ORC-1.7.0-SNAPSHOT (which you're definitively not using here). It only _possibly_ may be related to `SerializableConfiguration`, I am still debugging.

File: flink/src/main/java/org/apache/iceberg/flink/actions/FlinkActions.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  private static final Configuration CONFIG = new Configuration()
+      // disable classloader check as Avro may cache class/object in the serializers.
+      .set(CoreOptions.CHECK_LEAKED_CLASSLOADER, false);
+
+  private final StreamExecutionEnvironment env;
+
+  private FlinkActions(StreamExecutionEnvironment env) {
+    // register ClosureSerializer to prevent NPE when Kryo serialize lambda expr(e.g. HadoopFileIO.hadoopConf)
+    env.getConfig().registerTypeWithKryoSerializer(ClosureSerializer.Closure.class, ClosureSerializer.class);
```
Comment (Reo-LEI) [inline flink/src/main/java/org/apache/iceberg/flink/actions/FlinkActions.java:35]: I fix the NPE of kryo by register `ClosureSerializer` to flink, cuz Kryo serialize lambda use `ClosureSerializer.Closure`, but not register serialize for this class. I think that is a bug for Kryo or Flink. I hope that can help you @kbendick

**Pull request closed without merging**